File: plugin/python/larch/core/upgrade_larch.py
# pyright: reportUnknownVariableType=false, reportUnknownMemberType=false, reportUnknownArgumentType=false, reportUnusedCallResult=false, reportOptionalSubscript=false, reportOptionalMemberAccess=false, reportPossiblyUnboundVariable=false, reportUnnecessaryComparison=false, reportUnknownLambdaType=false, reportArgumentType=false
"""Python entrypoints for /upgrade-larch."""

from __future__ import annotations

import argparse
import json
import os
import re
import shutil
from contextlib import suppress
from pathlib import Path

from larch.core import config
from larch.core import logging_util
from larch.core import proc

LARCH_MARKETPLACE_SOURCE = "https://raw.githubusercontent.com/character-ai/larch/main/.claude-plugin/marketplace.json"
LARCH_SPARSE_DIRS = ".claude-plugin"
SAFE_VERSION_RE = re.compile(r"^[0-9]+(\.[0-9]+)*$")


def err(message: str = "") -> None:
    logging_util.BreadcrumbWriter().emit(message)


def is_safe_version(value: str | None) -> bool:
    return bool(value and SAFE_VERSION_RE.match(value))


def normalize_sparse_dirs() -> str:
    return "\n".join(sorted(part for part in LARCH_SPARSE_DIRS.split() if part))


def marketplace_clone_path(home: Path | None = None) -> Path | None:
    root: Path | None = (
        home or Path(os.environ["HOME"]) if os.environ.get("HOME") else None
    )
    return (
        root / ".claude/plugins/marketplaces/larch-local"
        if root and root.is_absolute()
        else None
    )


def release_step7_cache_parent(home: Path | None = None) -> Path | None:
    root: Path | None = (
        home or Path(os.environ["HOME"]) if os.environ.get("HOME") else None
    )
    return (
        root / ".claude/plugins/cache/larch-local/larch"
        if root and root.is_absolute()
        else None
    )


def _installed_larch_entries() -> list[dict[str, object]]:
    result = proc.run(["claude", "plugin", "list", "--json"])
    if result.returncode != 0:
        return []
    try:
        payload: object = json.loads(result.stdout)
    except json.JSONDecodeError:
        return []
    if not isinstance(payload, list):
        return []
    return [
        entry
        for entry in payload
        if isinstance(entry, dict) and entry.get("id") == "larch@larch-local"
    ]


def get_installed_larch_version() -> str:
    versions = {
        version
        for entry in _installed_larch_entries()
        if isinstance((version := entry.get("version")), str)
        and is_safe_version(version)
    }
    return versions.pop() if len(versions) == 1 else ""


def resolve_installed_larch_root(expected_version: str) -> Path | None:
    if not is_safe_version(expected_version):
        return None
    roots = {
        install_path
        for entry in _installed_larch_entries()
        if entry.get("version") == expected_version
        and isinstance((install_path := entry.get("installPath")), str)
        and install_path
    }
    if len(roots) != 1:
        return None
    root = Path(roots.pop())
    parent = release_step7_cache_parent()
    if (
        not root.is_absolute()
        or parent is None
        or root.is_symlink()
        or not root.is_dir()
    ):
        return None
    try:
        resolved = root.resolve(strict=True)
        resolved.relative_to(parent.resolve(strict=True))
    except (OSError, ValueError):
        return None
    manifest = resolved / ".claude-plugin/plugin.json"
    try:
        payload = json.loads(manifest.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    return (
        resolved
        if isinstance(payload, dict) and payload.get("version") == expected_version
        else None
    )


def is_cache_shaped_larch_root(root: str | Path | None) -> bool:
    if not root:
        return False
    path = Path(root)
    parent = release_step7_cache_parent()
    if parent is None:
        return False
    try:
        path.relative_to(parent)
    except ValueError:
        return False
    return path.is_dir() and is_safe_version(path.name)


def single_larch_cache_version_dir() -> Path | None:
    parent = release_step7_cache_parent()
    if parent is None or not parent.is_dir():
        return None
    dirs: list[Path] = [
        entry
        for entry in parent.iterdir()
        if entry.is_dir() and is_safe_version(entry.name)
    ]
    return dirs[0] if len(dirs) == 1 else None


def resolve_release_step7_root(current_version: str = "") -> Path | None:
    active = os.environ.get("CLAUDE_PLUGIN_ROOT", "")
    if is_cache_shaped_larch_root(active):
        return Path(active)
    parent = release_step7_cache_parent()
    if parent is None:
        return None
    installed_version = get_installed_larch_version()
    if is_safe_version(installed_version) and (parent / installed_version).is_dir():
        return parent / installed_version
    if is_safe_version(current_version):
        current_root = parent / current_version
        sole = single_larch_cache_version_dir()
        if current_root.is_dir() and sole == current_root:
            return current_root
    sole = single_larch_cache_version_dir()
    if is_safe_version(current_version) and sole == parent / current_version:
        return sole
    return None


def sparse_dirs_main(argv: list[str] | None = None) -> int:
    """Emit the normalized sparse checkout allowlist."""
    argparse.ArgumentParser(prog="cli.py upgrade-larch sparse-dirs").parse_args(argv)
    print(normalize_sparse_dirs())
    return 0


def release_step7_root_main(argv: list[str]) -> int:
    current_version = ""
    index = 0
    while index < len(argv):
        if argv[index] == "--current-version" and index + 1 < len(argv):
            current_version = argv[index + 1]
            index += 2
        elif not argv[index].startswith("-") and not current_version:
            current_version = argv[index]
            index += 1
        else:
            err(f"ERROR=Unknown argument: {argv[index]}")
            return 1
    root = resolve_release_step7_root(current_version)
    if root is None:
        err("ERROR=Unable to resolve larch cache root")
        return 1
    print(f"RESOLVED_ROOT={root}")
    return 0


def _marketplace_source_matches() -> bool:
    result = proc.run(["claude", "plugin", "marketplace", "list", "--json"])
    if result.returncode != 0:
        return False
    try:
        payload: object = json.loads(result.stdout)
    except json.JSONDecodeError:
        return False
    if not isinstance(payload, list):
        return False
    matches = [
        entry
        for entry in payload
        if isinstance(entry, dict) and entry.get("name") == "larch-local"
    ]
    return (
        len(matches) == 1
        and matches[0].get("source") == "url"
        and matches[0].get("url") == LARCH_MARKETPLACE_SOURCE
    )


def _gh_command(argv: list[str]) -> proc.CommandResult:
    from larch.git import gh  # noqa: PLC0415 - core leaf import stays function-local  # lint-layering: ok shared GitHub executable seam

    return gh.command(proc, argv)


def _get_stable_releases() -> list[str]:
    if shutil.which("gh") is None:
        err("Warning: gh is not available; upgrading without stable verification.")
        return []
    result = _gh_command(
        [
            "api",
            "--paginate",
            "repos/character-ai/larch/releases",
            "--jq",
            ".[] | select(.prerelease == false and .draft == false) | .tag_name",
        ]
    )
    if result.returncode != 0:
        err(
            f"Warning: failed to query GitHub stable releases via gh (exit {result.returncode}); upgrading without stable verification."
        )
        return []
    releases = [
        line.removeprefix("v")
        for line in result.stdout.splitlines()
        if is_safe_version(line.removeprefix("v"))
    ]
    if not releases:
        err(
            "Warning: gh returned no valid stable larch release tags; upgrading without stable verification."
        )
    return releases


def _recover_diagnostics() -> None:
    err("")
    err(
        "Recovery: retry /upgrade-larch. The running session and prior cache root remain usable."
    )
    err("If marketplace metadata is incomplete, run:")
    err(f"  claude plugin marketplace add {LARCH_MARKETPLACE_SOURCE}")
    err("  claude plugin install larch@larch-local")


def _refresh_marketplace() -> str:
    clone = marketplace_clone_path()
    if _marketplace_source_matches():
        err("Refreshing the runtime-only larch marketplace...")
        result = proc.run(["claude", "plugin", "marketplace", "update", "larch-local"])
        if result.returncode == 0:
            return "update"
        err("Marketplace refresh failed. The prior plugin cache root was not changed.")
        return ""
    err("Migrating the larch marketplace to the runtime-only remote source...")
    if clone is not None and clone.is_symlink():
        err("Marketplace migration refused a symlinked marketplace clone.")
        return ""
    removed = proc.run(["claude", "plugin", "marketplace", "remove", "larch-local"])
    if removed.returncode != 0:
        err(
            "Marketplace reconciliation stopped because the legacy registration could not be removed."
        )
        return ""
    if clone is not None and clone.exists():
        try:
            shutil.rmtree(clone)
        except OSError as exc:
            err(f"Warning: failed to remove marketplace clone '{clone}': {exc}")
            return ""
    add = proc.run(["claude", "plugin", "marketplace", "add", LARCH_MARKETPLACE_SOURCE])
    return "install" if add.returncode == 0 and _marketplace_source_matches() else ""


def _relay_output(result: proc.CommandResult) -> None:
    for line in result.stderr.splitlines():
        err(line)
    if result.stdout:
        print(result.stdout, end="" if result.stdout.endswith("\n") else "\n")


def _plugin_data_path() -> str:
    value = os.environ.get("CLAUDE_PLUGIN_DATA", "")
    return value if Path(value).is_absolute() else ""


def _preflight_release(*, version: str) -> bool:
    plugin_data = _plugin_data_path()
    script = Path(__file__).resolve().parents[3] / "scripts/larch.sh"
    if not plugin_data or not script.is_file() or script.is_symlink():
        err(
            "Upgrade preflight cannot resolve a safe bootstrap script and CLAUDE_PLUGIN_DATA path."
        )
        return False
    err(f"Preflighting immutable larch release v{version}...")
    result = proc.run(
        [str(script), "--preflight-release", version],
        timeout=600,
        env={**os.environ, "CLAUDE_PLUGIN_DATA": plugin_data},
    )
    _relay_output(result)
    return (
        result.returncode == 0
        and f"LARCH_PREFLIGHT_VERSION={version}" in result.stdout.splitlines()
    )


def _bootstrap_installed_root(*, root: Path, version: str) -> bool:
    plugin_data = _plugin_data_path()
    script = root / "scripts/larch.sh"
    binary = root / "bin/larch"
    if not plugin_data or not script.is_file() or script.is_symlink():
        err(
            "Installed plugin metadata resolved a root without a safe bootstrap script."
        )
        return False
    environment = {
        **os.environ,
        "CLAUDE_PLUGIN_ROOT": str(root),
        "CLAUDE_PLUGIN_DATA": plugin_data,
    }
    result = proc.run(
        [str(script), "bootstrap", "self-check"], timeout=600, env=environment
    )
    _relay_output(result)
    if (
        result.returncode != 0
        or not binary.is_file()
        or binary.is_symlink()
        or not os.access(binary, os.X_OK)
    ):
        return False
    direct = proc.run(
        [str(binary), "bootstrap", "self-check"], timeout=30, env=environment
    )
    if direct.returncode != 0 or direct.stdout != result.stdout:
        return False
    try:
        identity: object = json.loads(direct.stdout)
    except json.JSONDecodeError:
        return False
    return (
        isinstance(identity, dict)
        and identity.get("schema_version") == 1
        and identity.get("version") == version
    )


def _restore_operator_stdout() -> None:
    if os.environ.get(config.ENV_LARCH_QUIET_PID) == str(os.getpid()):
        with suppress(OSError):
            os.dup2(3, 1)


def run_main(argv: list[str]) -> int:  # noqa: PLR0911 - each upgrade gate fails closed with its own recovery boundary
    if argv:
        err(f"ERROR=Unknown argument: {argv[0]}")
        return 1
    logging_util.quiet_init(argv0="upgrade-larch.sh")
    _restore_operator_stdout()
    plugin_root = Path(os.environ.get("CLAUDE_PLUGIN_ROOT", Path.cwd())).resolve()
    installed_version = plugin_root.name
    latest = (
        os.environ.get("LARCH_EXPECTED_STABLE_VERSION", "")
        if is_safe_version(os.environ.get("LARCH_EXPECTED_STABLE_VERSION", ""))
        else ""
    )
    if not latest:
        releases = _get_stable_releases()
        latest = releases[0] if releases else ""
    current = get_installed_larch_version() or installed_version
    marketplace_will_reconcile = not _marketplace_source_matches()
    if (
        latest
        and current == latest
        and not marketplace_will_reconcile
        and (not is_cache_shaped_larch_root(plugin_root) or plugin_root.name == latest)
    ):
        current_root = resolve_installed_larch_root(current)
        if current_root is None or not _bootstrap_installed_root(
            root=current_root, version=current
        ):
            _recover_diagnostics()
            return 1
        err("")
        err(
            f"Already at latest stable larch release ({current}). Binary verification passed. No upgrade needed."
        )
        return 0
    if (
        latest
        and current == latest
        and is_cache_shaped_larch_root(plugin_root)
        and plugin_root.name != latest
    ):
        err("")
        err(
            f"Installed metadata is already at latest stable larch release ({current}), "
            f"but this Claude Code session is still running cached larch {plugin_root.name}. "
            "Refreshing the install and requiring restart...",
        )
    elif latest and current == latest and marketplace_will_reconcile:
        err("")
        err(
            f"Already at latest stable larch release ({current}), but the marketplace still uses the legacy source. Migrating it to the runtime-only source and reinstalling..."
        )
    elif latest:
        err(f"Upgrading larch from {installed_version} to {latest}...")
    else:
        err(
            "Latest stable release could not be determined. Upgrade stopped before changing plugin state."
        )
        _recover_diagnostics()
        return 1
    if not _preflight_release(version=latest):
        err("Upgrade stopped because stable release preflight failed.")
        _recover_diagnostics()
        return 1
    refresh_mode = _refresh_marketplace()
    if not refresh_mode:
        _recover_diagnostics()
        return 1
    err("Installing the preflighted larch plugin release...")
    command = [
        "claude",
        "plugin",
        "install" if refresh_mode == "install" else "update",
        "larch@larch-local",
    ]
    install = proc.run(command)
    if install.returncode != 0:
        err("Plugin install failed. The prior cache root was not modified.")
        _recover_diagnostics()
        return install.returncode or 1
    actual = get_installed_larch_version()
    verified = bool(latest and actual == latest)
    new_root = resolve_installed_larch_root(actual) if verified else None
    if (
        not verified
        or new_root is None
        or not _bootstrap_installed_root(root=new_root, version=actual)
    ):
        err(
            f"Upgrade incomplete: expected plugin and binary version {latest} in the newly installed cache root."
        )
        _recover_diagnostics()
        return 1
    if marketplace_will_reconcile:
        err(
            f"LARCH_MARKETPLACE_RECONCILED={'true' if _marketplace_source_matches() else 'false'}"
        )
    err("LARCH_RESTART_REQUIRED=true")
    if actual != current:
        err("LARCH_NEW_VERSION_INSTALLED=true")
    err("")
    err("Installed larch plugin version:")
    listed = proc.run(["claude", "plugin", "list"])
    if listed.stdout:
        print(listed.stdout, end="")
    err("")
    err("Upgrade complete. Restart Claude Code to apply the new version.")
    return 0

File: crates/larch-cli/tests/upgrade_larch.rs
#![cfg(unix)]

use assert_cmd::Command;
use predicates::prelude::*;
use std::{
    fs,
    os::unix::fs::PermissionsExt,
    path::{Path, PathBuf},
};
use tempfile::TempDir;

const SOURCE: &str =
    "https://raw.githubusercontent.com/character-ai/larch/main/.claude-plugin/marketplace.json";

struct Harness {
    _temp: TempDir,
    home: PathBuf,
    data: PathBuf,
    fake_bin: PathBuf,
    old_root: PathBuf,
    new_root: PathBuf,
    state: PathBuf,
    log: PathBuf,
}

impl Harness {
    fn new() -> Self {
        let temp = tempfile::tempdir().expect("tempdir");
        let home = temp.path().join("home");
        let data = temp.path().join("data");
        let fake_bin = temp.path().join("fake-bin");
        let state = temp.path().join("state");
        let log = temp.path().join("calls.log");
        fs::create_dir_all(&fake_bin).expect("fake bin");
        fs::create_dir_all(&data).expect("plugin data");
        fs::create_dir_all(&state).expect("state");
        fs::write(state.join("installed"), "1.0.0").expect("installed state");
        fs::write(state.join("desired"), "2.0.0").expect("desired state");
        fs::write(state.join("marketplace"), "remote").expect("marketplace state");
        let cache = home.join(".claude/plugins/cache/larch-local/larch");
        let old_root = cache.join("1.0.0");
        let new_root = cache.join("2.0.0");
        install_root(&old_root, "1.0.0", &state, &log);
        install_root(&new_root, "2.0.0", &state, &log);
        write_executable(
            &fake_bin.join("claude"),
            &claude_script(&state, &log, &old_root, &new_root),
        );
        Self {
            _temp: temp,
            home,
            data,
            fake_bin,
            old_root,
            new_root,
            state,
            log,
        }
    }

    fn command(&self) -> Command {
        let coverage_profile = std::env::var_os("LLVM_PROFILE_FILE");
        let mut command = Command::cargo_bin("larch").expect("larch binary");
        command
            .env_clear()
            .env("HOME", &self.home)
            .env("PATH", format!("{}:/bin", self.fake_bin.display()))
            .env("CLAUDE_PLUGIN_DATA", &self.data)
            .env("CLAUDE_PLUGIN_ROOT", &self.old_root)
            .env("LARCH_EXPECTED_STABLE_VERSION", "2.0.0");
        if let Some(profile) = coverage_profile {
            command.env("LLVM_PROFILE_FILE", profile);
        }
        command
    }

    fn set(&self, name: &str, value: &str) {
        fs::write(self.state.join(name), value).expect("state write");
    }

    fn flag(&self, name: &str) {
        fs::write(self.state.join(name), "1").expect("flag");
    }

    fn clear(&self, name: &str) {
        let _ = fs::remove_file(self.state.join(name));
    }
}

#[test]
fn auxiliary_commands_preserve_their_machine_output() {
    let harness = Harness::new();
    harness
        .command()
        .arg("upgrade-larch")
        .arg("sparse-dirs")
        .assert()
        .success()
        .stdout(".claude-plugin\n");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("release-step7-root")
        .assert()
        .success()
        .stdout(format!("RESOLVED_ROOT={}\n", harness.old_root.display()));

    harness
        .command()
        .args([
            "upgrade-larch",
            "release-step7-root",
            "1.2.3",
            "--current-version",
            "1.2.3",
        ])
        .assert()
        .failure()
        .stderr(predicates::str::contains("an argument cannot be used with"));
}

#[test]
fn no_op_repairs_and_verifies_the_binary_without_python() {
    let harness = Harness::new();
    harness.set("installed", "2.0.0");
    fs::remove_file(harness.new_root.join("bin/larch")).expect("remove broken binary");
    harness
        .command()
        .env("CLAUDE_PLUGIN_ROOT", &harness.new_root)
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .success()
        .stderr(predicate::str::contains(
            "Binary verification passed. No upgrade needed.",
        ));
    assert!(harness.new_root.join("bin/larch").is_file());
    assert!(
        !fs::read_to_string(&harness.log)
            .expect("log")
            .contains("python")
    );
}

#[test]
fn ordinary_upgrade_preflights_then_preserves_the_active_old_root() {
    let harness = Harness::new();
    let marker = harness.old_root.join("active-session");
    fs::write(&marker, "old").expect("marker");
    harness
        .command()
        .env("GH_TOKEN", "must-not-be-forwarded")
        .env("GITHUB_TOKEN", "must-not-be-forwarded")
        .env("GH_CONFIG_DIR", harness.home.join(".config/gh"))
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .success()
        .stderr(
            predicate::str::contains("LARCH_RESTART_REQUIRED=true")
                .and(predicate::str::contains("LARCH_NEW_VERSION_INSTALLED=true")),
        );
    assert_eq!(fs::read_to_string(marker).expect("old marker"), "old");
    let log = fs::read_to_string(&harness.log).expect("log");
    let preflight = log
        .find("bootstrap --preflight-release 2.0.0")
        .expect("preflight");
    let refresh = log
        .find("claude plugin marketplace update larch-local")
        .expect("refresh");
    assert!(preflight < refresh);
    assert!(!log.contains("bootstrap-gh-token-present"));
    assert!(log.contains("bootstrap-gh-config-present"));
}

#[test]
fn release_step7_can_keep_execution_and_installed_roots_separate() {
    let harness = Harness::new();
    harness
        .command()
        .env("CLAUDE_PLUGIN_ROOT", &harness.new_root)
        .args(["upgrade-larch", "run", "--plugin-root"])
        .arg(&harness.old_root)
        .assert()
        .success()
        .stderr(predicate::str::contains("LARCH_NEW_VERSION_INSTALLED=true"));
}

#[test]
fn preflight_uses_the_driver_root_when_the_install_target_predates_the_flag() {
    let harness = Harness::new();
    // Simulate upgrading from a version whose larch.sh predates
    // `--preflight-release` (e.g. 53.x), which errors on the flag. The preflight
    // must still succeed by running the driver's (CLAUDE_PLUGIN_ROOT) larch.sh,
    // not the install target's.
    write_executable(
        &harness.old_root.join("scripts/larch.sh"),
        "#!/bin/sh\ncase \"$1\" in\n--preflight-release) echo 'unexpected argument' >&2; exit 1 ;;\nesac\n",
    );
    harness
        .command()
        .env("CLAUDE_PLUGIN_ROOT", &harness.new_root)
        .env("LARCH_EXPECTED_STABLE_VERSION", "2.0.0")
        .args(["upgrade-larch", "run", "--plugin-root"])
        .arg(&harness.old_root)
        .assert()
        .success()
        .stderr(predicate::str::contains("LARCH_NEW_VERSION_INSTALLED=true"));
}

#[test]
fn active_old_session_is_refreshed_without_deleting_its_cache_root() {
    let harness = Harness::new();
    harness.set("installed", "2.0.0");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .success()
        .stderr(
            predicate::str::contains("still running cached larch 1.0.0")
                .and(predicate::str::contains("LARCH_RESTART_REQUIRED=true")),
        );
    assert!(harness.old_root.join("bin/larch").is_file());
    assert!(harness.new_root.join("bin/larch").is_file());
}

#[test]
fn marketplace_migration_removes_only_the_legacy_clone_and_reinstalls() {
    let harness = Harness::new();
    harness.set("marketplace", "legacy");
    let clone = harness
        .home
        .join(".claude/plugins/marketplaces/larch-local");
    fs::create_dir_all(&clone).expect("clone");
    fs::write(clone.join("legacy"), "legacy").expect("legacy marker");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .success()
        .stderr(predicate::str::contains(
            "LARCH_MARKETPLACE_RECONCILED=true",
        ));
    assert!(!clone.exists());
    assert!(harness.old_root.is_dir());
    let log = fs::read_to_string(&harness.log).expect("log");
    assert!(log.contains(&format!("claude plugin marketplace add {SOURCE}")));
    assert!(log.contains("claude plugin install larch@larch-local"));
}

#[test]
fn ambiguous_installed_roots_fail_closed_with_retry_guidance() {
    let harness = Harness::new();
    harness.set("installed", "2.0.0");
    harness.flag("ambiguous");
    harness
        .command()
        .env("CLAUDE_PLUGIN_ROOT", &harness.new_root)
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .failure()
        .stderr(predicate::str::contains("Recovery: retry /upgrade-larch"));
}

#[test]
fn interrupted_preflight_is_retryable_and_does_not_mutate_plugin_state() {
    let harness = Harness::new();
    harness.flag("fail_preflight");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .failure()
        .stderr(predicate::str::contains("stable release preflight failed"));
    assert_eq!(
        fs::read_to_string(harness.state.join("installed")).expect("state"),
        "1.0.0"
    );
    harness.clear("fail_preflight");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .success();
}

#[test]
fn missing_plugin_data_fails_closed_before_any_mutation_with_the_remedy() {
    let harness = Harness::new();
    harness
        .command()
        .env_remove("CLAUDE_PLUGIN_DATA")
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .failure()
        .stderr(
            predicate::str::contains("CLAUDE_PLUGIN_DATA is required")
                .and(predicate::str::contains("docs/installation-and-setup.md")),
        );
    let log = fs::read_to_string(&harness.log).expect("log");
    assert!(!log.contains("bootstrap --preflight-release"));
    assert_eq!(
        fs::read_to_string(harness.state.join("installed")).expect("installed state"),
        "1.0.0"
    );
    assert_eq!(
        fs::read_to_string(harness.state.join("marketplace")).expect("marketplace state"),
        "remote"
    );
}

#[test]
fn failed_refresh_and_failed_install_leave_the_prior_root_usable() {
    let harness = Harness::new();
    harness.flag("fail_refresh");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .failure()
        .stderr(predicate::str::contains(
            "prior plugin cache root was not changed",
        ));
    assert!(harness.old_root.join("bin/larch").is_file());
    harness.clear("fail_refresh");
    harness.flag("fail_install");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .code(7)
        .stderr(predicate::str::contains("Plugin install failed"));
    assert!(harness.old_root.join("bin/larch").is_file());
}

#[test]
fn failed_new_root_bootstrap_preserves_rollback_and_retry_state() {
    let harness = Harness::new();
    harness.flag("fail_bootstrap_2.0.0");
    let marker = harness.old_root.join("rollback-marker");
    fs::write(&marker, "old").expect("marker");
    harness
        .command()
        .arg("upgrade-larch")
        .arg("run")
        .assert()
        .failure()
        .stderr(predicate::str::contains("Upgrade incomplete"));
    assert_eq!(fs::read_to_string(marker).expect("marker"), "old");
    assert!(harness.new_root.is_dir());
}

fn install_root(root: &Path, version: &str, state: &Path, log: &Path) {
    fs::create_dir_all(root.join(".claude-plugin")).expect("manifest dir");
    fs::create_dir_all(root.join("scripts")).expect("scripts dir");
    fs::create_dir_all(root.join("bin")).expect("bin dir");
    fs::write(
        root.join(".claude-plugin/plugin.json"),
        format!(r#"{{"version":"{version}"}}"#),
    )
    .expect("manifest");
    let binary = identity_script(version, log);
    write_executable(&root.join("bin-template"), &binary);
    write_executable(&root.join("bin/larch"), &binary);
    write_executable(
        &root.join("scripts/larch.sh"),
        &bootstrap_script(root, version, state, log),
    );
}

fn identity_script(version: &str, log: &Path) -> String {
    format!(
        "#!/bin/sh\nprintf 'binary %s\\n' \"$*\" >> '{}'\nprintf '{{\"schema_version\":1,\"version\":\"{version}\",\"target\":\"test-target\"}}\\n'\n",
        log.display()
    )
}

fn bootstrap_script(root: &Path, version: &str, state: &Path, log: &Path) -> String {
    format!(
        r#"#!/bin/sh
printf 'bootstrap %s\n' "$*" >> '{log}'
[ -z "${{GH_TOKEN:-}}" ] || printf 'bootstrap-gh-token-present\n' >> '{log}'
[ -z "${{GITHUB_TOKEN:-}}" ] || printf 'bootstrap-gh-token-present\n' >> '{log}'
[ -z "${{GH_CONFIG_DIR:-}}" ] || printf 'bootstrap-gh-config-present\n' >> '{log}'
case "${{1:-}}" in
  --latest-stable-version) printf 'LARCH_STABLE_VERSION=2.0.0\n' ;;
  --preflight-release)
    [ ! -f '{state}/fail_preflight' ] || exit 1
    printf 'LARCH_PREFLIGHT_VERSION=%s\n' "$2"
    ;;
  bootstrap)
    [ ! -f '{state}/fail_bootstrap_{version}' ] || exit 1
    if [ ! -x '{root}/bin/larch' ]; then
      /bin/cp '{root}/bin-template' '{root}/bin/larch'
      /bin/chmod 755 '{root}/bin/larch'
    fi
    exec '{root}/bin/larch' "$@"
    ;;
esac
"#,
        log = log.display(),
        state = state.display(),
        root = root.display()
    )
}

fn claude_script(state: &Path, log: &Path, old_root: &Path, new_root: &Path) -> String {
    format!(
        r#"#!/bin/sh
printf 'claude %s\n' "$*" >> '{log}'
installed="$(/bin/cat '{state}/installed')"
case "$*" in
  'plugin list --json')
    if [ "$installed" = 2.0.0 ]; then root='{new_root}'; else root='{old_root}'; fi
    if [ -f '{state}/ambiguous' ]; then
      printf '[{{"id":"larch@larch-local","version":"%s","installPath":"%s"}},{{"id":"larch@larch-local","version":"%s","installPath":"%s/other"}}]\n' "$installed" "$root" "$installed" "$root"
    else
      printf '[{{"id":"larch@larch-local","version":"%s","installPath":"%s"}}]\n' "$installed" "$root"
    fi
    ;;
  'plugin marketplace list --json')
    if [ "$(/bin/cat '{state}/marketplace')" = remote ]; then
      printf '[{{"name":"larch-local","source":"url","url":"{source}"}}]\n'
    else
      printf '[{{"name":"larch-local","source":"github","url":"legacy"}}]\n'
    fi
    ;;
  'plugin marketplace update larch-local') [ ! -f '{state}/fail_refresh' ] || exit 1 ;;
  'plugin marketplace remove larch-local') : ;;
  'plugin marketplace add {source}') printf remote > '{state}/marketplace' ;;
  'plugin update larch@larch-local'|'plugin install larch@larch-local')
    [ ! -f '{state}/fail_install' ] || exit 7
    /bin/cat '{state}/desired' > '{state}/installed'
    ;;
  'plugin list') printf 'larch@larch-local %s\n' "$installed" ;;
esac
"#,
        log = log.display(),
        state = state.display(),
        old_root = old_root.display(),
        new_root = new_root.display(),
        source = SOURCE
    )
}

fn write_executable(path: &Path, body: &str) {
    fs::write(path, body).expect("script write");
    let mut permissions = fs::metadata(path).expect("script metadata").permissions();
    permissions.set_mode(0o755);
    fs::set_permissions(path, permissions).expect("script permissions");
}
